### PR TITLE
DDF-3297 Adds security manager to container

### PR DESCRIPTION
**Pull request by coyotesqrl**
#### What does this PR do?
Adds ProGrade security manager and policy file to container.

#### Who is reviewing it? 
@paouelle @Lambeaux @SmithJosh @garrettfreibott @ryeats 

(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)
#### Select relevant component teams: 
@codice/security 

#### Choose 2 committers to review/merge the PR. 
(please choose ONLY two committers from below, delete the rest)
@clockard
@lessarderic
@stustison

#### How should this be tested? (List steps with links to updated documentation)
Full regression test of DDF. This will potentially impact downstream projects; they will need to add/update policy files and `setenv` scripts accordingly.

#### Any background context you want to provide?
N/A

#### What are the relevant tickets?
[DDF-3297](https://codice.atlassian.net/browse/DDF-3297)

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests
#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: distribution/ddf-common/src/main/resources/etc/default.policy [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+// todo symlinks
+
+deny {
+    // Read, write, or execute any file
+    permission java.io.FilePermission "<<ALL FILES>>", "read, write, execute";
+
+    // Deny deletion of the security policy file through any Java code
+    permission java.io.FilePermission "${ddf.home}${/}etc${/}default.policy", "delete";
+
+    permission java.util.PropertyPermission "javax.net.ssl.*", "read, write";
```
Comment (paouelle) [inline distribution/ddf-common/src/main/resources/etc/default.policy]: migration export/import needs read access to:

javax.net.ssl.keyStore
javax.net.ssl.trustStore

File: distribution/ddf-common/src/main/resources/etc/default.policy [hunk 2]
```diff
@@ -0,0 +1,120 @@
+priority "grant";
+// todo symlinks
+
+deny {
+    // Read, write, or execute any file
+    permission java.io.FilePermission "<<ALL FILES>>", "read, write, execute";
+
+    // Deny deletion of the security policy file through any Java code
+    permission java.io.FilePermission "${ddf.home}${/}etc${/}default.policy", "delete";
```
Comment (paouelle) [inline distribution/ddf-common/src/main/resources/etc/default.policy]: Should we block write too?
import will need write access though to restore that file.

File: distribution/ddf-common/src/main/resources/etc/default.policy [hunk 3]
```diff
@@ -0,0 +1,120 @@
+priority "grant";
+// todo symlinks
+
+deny {
+    // Read, write, or execute any file
+    permission java.io.FilePermission "<<ALL FILES>>", "read, write, execute";
```
Comment (paouelle) [inline distribution/ddf-common/src/main/resources/etc/default.policy]: Should also be blocking delete?

File: distribution/ddf-common/src/main/resources/etc/default.policy [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    permission java.io.FilePermission "${java.home}${/}lib${/}-", "read";
+
+    // System libraries
+    permission java.io.FilePermission "/dev/urandom", "read";
+    permission java.io.FilePermission "/lib", "read";
+    permission java.io.FilePermission "/proc/self/exe", "read";
+    permission java.io.FilePermission "/usr/lib", "read";
+
+
+    // Distribution File Permissions
```
Comment (figliold) [inline distribution/ddf-common/src/main/resources/etc/default.policy]: @paouelle Wouldn't the export/import need access to the directory the exported zip was saved in?

Comment (paouelle) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: and a lot more. I am compiling the required changes in a separate spike.

File: distribution/ddf-common/src/main/resources/bin/setenv [hunk 5]
```diff
@@ -64,5 +64,14 @@ export EXTRA_JAVA_OPTS
...
 
-export KARAF_OPTS="-Dfile.encoding=UTF8 -Dddf.home=$DDF_HOME"
+#
+# Admins can uncomment the following line and comment out the other definition of
+# the KARAF_SYSTEM_OPTS variable in order to determine any missing security permissions
+# should they install third-party bundles requiring additional access.
+#
+# export KARAF_SYSTEM_OPTS="-Dprograde.generated.policy=$DDF_HOME/generated.policy -Dprograde.use.own.policy=true -Djava.security.manager=net.sourceforge.prograde.sm.PolicyFileGeneratorJSM -Djava.security.policy==$DDF_HOME/etc/default.policy -DproGrade.getPermissions.override=sun.rmi.server.LoaderHandler:loadClass,org.apache.jasper.compiler.JspRuntimeContext:initSecurity"
+
+export KARAF_SYSTEM_OPTS="-Djava.security.manager=net.sourceforge.prograde.sm.ProGradeJSM -Djava.security.policy==$DDF_HOME/etc/default.policy -DproGrade.getPermissions.override=sun.rmi.server.LoaderHandler:loadClass,org.apache.jasper.compiler.JspRuntimeContext:initSecurity"
```
Comment (coyotesqrl) [inline distribution/ddf-common/src/main/resources/bin/setenv]: Using `KARAF_SYSTEM_OPTS` here and in the Windows script in order to provide differentiation between `client` and `ddf` scripts. There are other ways we could handle the different needs - the client script does _not_ need a security manager - but they would require managing more files separate from the Karaf distro.

File: distribution/ddf-common/src/main/resources/etc/default.policy (hunk 3 above)
Comment (coyotesqrl) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: It should, and future iterations (likely beyond 2.11.0) will. Could also consider blocking `readLink`, but that's of lower concern.

File: distribution/ddf-common/src/main/resources/etc/default.policy (hunk 2 above)
Comment (coyotesqrl) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: The permission line above this is already globally blocking read, write, and execute.

File: distribution/ddf-common/src/main/resources/etc/default.policy (hunk 1 above)
Comment (coyotesqrl) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: Let me know which bundle(s) need those permissions and I'll grant them.

Comment (paouelle) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: It's a place holder for now. I will compile all the requirements for export/import separatly

File: distribution/ddf-common/src/main/resources/etc/default.policy [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    // Java libraries
+    permission java.io.FilePermission "${java.home}${/}release", "read";
+    permission java.io.FilePermission "${java.home}${/}lib${/}-", "read";
+
+    // System libraries
+    permission java.io.FilePermission "/dev/urandom", "read";
+    permission java.io.FilePermission "/lib", "read";
+    permission java.io.FilePermission "/proc/self/exe", "read";
+    permission java.io.FilePermission "/usr/lib", "read";
+
```
Comment (lessarderic) [inline distribution/ddf-common/src/main/resources/etc/default.policy]: ❓ How about temporary folders, e.g., `/tmp`, `/var/tmp`, `C:\TEMP`, etc.? The `FileBackedOutputStream` class (which is used in quite a few places in our code) needs to be able to create temporary files: https://github.com/google/guava/blob/master/guava/src/com/google/common/io/FileBackedOutputStream.java#L195.

Comment (coyotesqrl) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: We will need to test this across platforms, but it is almost definitely broken right now.

File: distribution/ddf-common/src/main/resources/etc/default.policy (hunk 2 above)
Comment (coyotesqrl) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: We do grant many `codeBase`s access to `etc`, but it's _mostly_ restricted. We will lock it down more, but this should be sufficiently safe for now.

**Comment by paouelle** [comment]
We need to update the documentation with respect to service creation since the changes to setenv won't propagate there. We should also re-test the security manager when running as a service

File: distribution/ddf-common/src/main/resources/etc/default.policy (hunk 1 above)
Comment (paouelle) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: Can you add the following permissions for export/import?

//
// DDF Migration Framework and Config Admin Migratable
//
grant
  codeBase "file:${ddf.home.policy}system/ddf/platform/platform-migratable/-" {
    // some permissions might be repeated here to ensure that they are granted to the migration
    // framework if we ever tighten the permissions above

    permission java.io.FilePermission "<<ALL FILES>>", "read, readlink";
    permission java.io.FilePermission "${ddf.home}", "read, write";

    permission java.util.PropertyPermission "*", "read";
    permission java.util.PropertyPermission "ddf.home", "read";
    permission java.util.PropertyPermission "karaf.restart.jvm", "write";
};

//
// DDF Platform Migratable
//
grant
  codeBase "file:${ddf.home.policy}system/ddf/platform/platform-migratable/-" {
    // some permissions might be repeated here to ensure that they are granted to the migration
    // framework if we ever tighten the permissions above

    permission java.io.FilePermission "${ddf.home}${/}etc${/}ws-security", "read";
    permission java.io.FilePermission "${ddf.home}${/}etc${/}ws-security${/}-", "read, delete";

    permission java.util.PropertyPermission "javax.net.ssl.keyStore", "read";
    permission java.util.PropertyPermission "javax.net.ssl.trustStore", "read";
};

//
// DDF Security Migratable
//
grant
  codeBase "file:${ddf.home.policy}system/ddf/security/migration/security-migratable/-" {
    // some permissions might be repeated here to ensure that they are granted to the migration
    // framework if we ever tighten the permissions above

    permission java.io.FilePermission "${ddf.home}${/}etc${/}pdp", "read";
    permission java.io.FilePermission "${ddf.home}${/}etc${/}pdp{/}-", "read, delete";
};

Comment (coyotesqrl) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: Will do, @paouelle

**Comment by coyotesqrl** [comment]
build now

**Comment by rzwiefel** [comment]
@coyotesqrl This PR is marked with `2.11.0` but is pointed into codice:master.. Is this PR supposed to be opened against `codice:2.11.x` instead?

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
@rzwiefel It should have the milestone removed. It was intended to be merged prior to the branch; a backport will be put up after this is approved and merged.

File: distribution/ddf-common/src/main/resources/etc/default.policy (hunk 6 above)
Comment (lessarderic) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: Even though things worked because `FileBackedOutStream` defaults to a directory under `ddf.home`, we may still want to grant read access to the `java.io.tmpdir` system property and allow read/write access to the directory it points to. This would prevent any unexpected issues if a third-party library decides to create some temporary files there.

**Comment by coyotesqrl** [comment]
@lessarderic I should have just looked to begin with. The `karaf` and `karaf.bat` scripts set `java.io.tmpdir` to `${KARAF_DATA}/tmp`. I'll explicitly add that path to the global grants, but it should already be covered.

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
Build now

**Comment by coyotesqrl** [comment]
Build now

File: distribution/ddf-common/src/main/resources/etc/default.policy (hunk 6 above)
Comment (paouelle) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: don't forget delete too

Comment (coyotesqrl) [reply distribution/ddf-common/src/main/resources/etc/default.policy]: Thanks, @paouelle. Fixed.

File: distribution/ddf-common/src/main/resources/etc/default.policy [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    permission java.lang.RuntimePermission "createClassLoader";
+};
+
+//
+// DDF Libraries have broad file permissions within the install directory
+//
+grant
+  codeBase "file:${ddf.home.policy}system/ddf/-" {
+    permission java.io.FilePermission "${ddf.home}", "read, write";
+    permission java.io.FilePermission "${ddf.home}${/}etc${/}-", "read, write";
```
Comment (paouelle) [inline distribution/ddf-common/src/main/resources/etc/default.policy]: Felix config installer (and Steve's replacement) will also need delete for .cfg and .config files located under etc (not recursive)

File: distribution/ddf-common/src/main/resources/etc/default.policy [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+    // Java libraries
+    permission java.io.FilePermission "${java.home}${/}release", "read";
+    permission java.io.FilePermission "${java.home}${/}lib${/}-", "read";
+
+    // System libraries
+    permission java.io.FilePermission "/dev/urandom", "read";
+    permission java.io.FilePermission "/lib", "read";
+    permission java.io.FilePermission "/proc/self/exe", "read";
+    permission java.io.FilePermission "/usr/lib", "read";
```
Comment (stustison) [inline distribution/ddf-common/src/main/resources/etc/default.policy]: Are there any Windows equivalents for these that we need to worry about?

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java [hunk 9]
```diff
@@ -585,7 +585,12 @@ private static Integer findPortNumber(Integer portToTry) {
         vmOption("-Xmx2048M"),
         // avoid integration tests stealing focus on OS X
         vmOption("-Djava.awt.headless=true"),
-        vmOption("-Dfile.encoding=UTF8"));
+        vmOption("-Dfile.encoding=UTF8"),
+        vmOption("-Djava.security.manager=net.sourceforge.prograde.sm.ProGradeJSM"),
+        vmOption("-Dpolicy.provider=net.sourceforge.prograde.policy.ProGradePolicy"),
+        vmOption("-Djava.security.policy==etc/all.policy"),
```
Comment (stustison) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java]: I'd feel better if this was default.policy + extra permissions for itests

File: distribution/ddf-common/src/main/resources/security/default.policy [hunk 10]
```diff
@@ -6,7 +6,7 @@ deny {
     permission java.io.FilePermission "<<ALL FILES>>", "read, write, execute";
 
     // Deny deletion of the security policy file through any Java code
```
Comment (paouelle) [inline distribution/ddf-common/src/main/resources/security/default.policy]: ✏️  might want to update the comment to reflect what we are denying

**Comment by coyotesqrl** [comment]
build now

**Comment by paouelle** [comment]
re-tested latest code (i.e. default.policy) with PR 2409 (i.e. export/import changes for security manager) and it is 👍

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 11]
```diff
@@ -703,54 +703,6 @@ public void testFederatedRetrieveProductInvalidResourceUrl() throws Exception {
         .body(containsString("Unable to transform Metacard."));
   }
 
-  /**
```
Comment (coyotesqrl) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: The two tests removed from this class no longer make sense. They fail early because of permissions that protect the system and there's no reason for the tests to continue.

**Comment by coyotesqrl** [comment]
build now

File: distribution/ddf-common/src/main/resources/bin/setenv [hunk 12]
```diff
@@ -64,5 +64,18 @@ export EXTRA_JAVA_OPTS
 # export KARAF_DATA # Karaf data folder
 # export KARAF_BASE # Karaf base folder
 
-export KARAF_OPTS="-Dfile.encoding=UTF8 -Dddf.home=$DDF_HOME"
+#
+# Admins can uncomment the following line and comment out the other definition of
+# the KARAF_SYSTEM_OPTS variable in order to determine any missing security permissions
+# should they install third-party bundles requiring additional access.
```
Comment (paouelle) [inline distribution/ddf-common/src/main/resources/bin/setenv]: ❓ If uncommenting this line grants everything by default, we might want to add a comment to that regard

Comment (coyotesqrl) [reply distribution/ddf-common/src/main/resources/bin/setenv]: Good point. It is a risk if left on. I'll expand the comment here and in the windows script.

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
build now

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java [hunk 13]
```diff
@@ -794,4 +821,33 @@ protected boolean isCatalogEmpty() {
       return false;
     }
   }
+
+  /**
+   * Helper Option class to allow interpolation of {@code karaf.home} directory based on the
+   * provided {@link #UNPACK_DIRECTORY}.
+   */
+  static class HomeAwareVmOption extends VMOption {
```
Comment (tbatie) [inline distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java]: Can we leave a comment mentioning why this was made and the edge case it's addressing? I don't anyone else should really ever be using this.

**Comment by paouelle** [comment]
❓ Should we undo the PR that disabled SSH as part of this one?

**Comment by paouelle** [comment]
❗️ The security manager stuff will be problematic when running as a service. Checkout [My application is throwing security exceptions when I run with the Wrapper.](https://wrapper.tanukisoftware.com/doc/english/faq.html#5)

File: distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java (hunk 13 above)
Comment (coyotesqrl) [reply distribution/test/itests/test-itests-common/src/main/java/org/codice/ddf/itests/common/AbstractIntegrationTest.java]: Considering it's itest code and does have a short javadoc, I don't know we need too much more. I doubt anyone would duplicate this approach.

**Comment by coyotesqrl** [comment]
@paouelle The SSH change is going to be separate work that follows this. It will also include shutting off permissions to some Felix console commands.

**Comment by coyotesqrl** [comment]
@paouelle We'll need to run as a service on Windows and Linux and determine how the security manager interplays. I would probably be fine granting the wrapper all permissions such that our codeBases would get the intersection of their restricted permissions and all permissions.

**Comment by paouelle** [comment]
@coyotesqrl the code base for the wrapper is lib/wrapper.

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
build now

**Comment by oconnormi** [comment]
Build Now

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
build now

**Comment by coyotesqrl** [comment]
build now

**Comment by oconnormi** [comment]
:tada:
